app/Database/CourseInsertion.hs

```diff
@@ -39,10 +43,28 @@ saveGraphJSON jsonStr nameStr = do
                 insertMany_ $ map (\path -> path {pathGraph = gId}) paths
             return $ toResponse $ ("Success" :: String)
 
--- | Inserts course into the Courses table.
+
+-- | Check for duplicate key, then if unique, insert course into Courses table
+--safeInsertCourse :: Course -> ReaderT SqlBackend m (Maybe Course) -- this screws up 'insert'
+--safeInsertCourse :: Course -> ReaderT SqlBackend m () -- this screws up 'insert'
+safeInsertCourse course = do
+    maybeCourse <- selectFirst [CoursesCode ==. (name course)] [] -- maybeCourse :: [Entity Courses] ?
```

Comment:
Look at the documentation for selectFirst (https://hackage.haskell.org/package/persistent-2.2.4/docs/Database-Persist-Class.html#v:selectFirst). The `maybeCourse` variable has type which is equal to the last part of the returned type, which is `Maybe ____` (in this case, `Maybe Courses`).

But you seem to be using already using that type based on what you do in the next three lines.